---
# Seminar 1 - Grundlagen
---

### Hinweis: Speichern eines Notebooks immer über: STRG+S/ CMD+S oder das Disketten-Symbol. Die Option File --> Save funktioniert bei Notebooks nicht!!


## 1. Podcast-Sendezeit

Sie haben es geschafft und konnten sich einen der begehrten Praktikumsplätze im Team des Podcast-Newcomers "OR" ergattern. Ihre erste Aufgabe besteht darin, herauszufinden, wie viele Sendeminuten der Podcast wöchentlich hat.

1. Eine Podcastfolge dauert im Durchschnitt 70 Minuten. Dies soll aber variabel anpassbar sein. (Variable)
2. Zusätzlich werden pro Folge 3 Werbeblöcke à 3 Minuten eingespielt.
3. Wöchentlich werden 2 Podcasts veröffentlicht.
4. Die Redaktionschefin bittet Sie, das Ergebnis in Stunden gerundet auf zwei Nachkommastellen auszugeben.
5. Des Weiteren soll nicht nur einfach eine Zahl ausgegeben werden, sondern ein Satz: "Bei einer Podcastlänge von XX Minuten ergeben sich Y.YY Stunden Sendezeit pro Woche."

### a.)

Schreiben Sie bitte ein Skript, welches die Berechnung übernimmt und geben Sie das Ergebnis aus.


In [4]:
duration = 70

airTimeperWeek = round((duration + 3*3)*2/60, 2)

print('Bei einer Podcastlänge von ' + str(duration) + ' Minuten ergeben sich ' + str(airTimeperWeek) + ' Stunden Sendezeit')

Bei einer Podcastlänge von 70 Minuten ergeben sich 2.63 Stunden Sendezeit


In [3]:
print('播客长度为' + str(duration) + '每周有' + str(airTimeperWeek) + '小时的广播时间。')

播客长度为70每周有2.63小时的广播时间。


In [2]:
duration = 70
airTimeperWeek = round((duration + 3*3) * 2 / 60, 2)

print(f'播客长度为{duration}分钟，每周有{airTimeperWeek}小时的广播时间。')

播客长度为70分钟，每周有2.63小时的广播时间。


### b.)

Informieren Sie sich bitte im Internet, welche weiteren Möglichkeiten es gibt, Zeichenketten für eine Ausgabe zu formatieren.

**Tipp:**
<details>

Schreiben Sie den Ausgabebefehl von Aufgabe a) als "".format() und f""-Ausgabe.
</details>

In [6]:
print(f'Bei einer Podcastlänge von {duration} Minuten ergeben sich {airTimeperWeek} Stunden Sendezeit.')

Bei einer Podcastlänge von 70 Minuten ergeben sich 2.63 Stunden Sendezeit.


Erwarteter Output:

    Bei einer Podcastlänge von 70 Minuten ergeben sich ...

## 2. Chaos in der Mailingliste

Die Hörer des Podcasts haben die Möglichkeit, sich für einen Newsletter anzumelden. So erfahren sie schnellstmöglich, wann es eine neue Podcast-Folge gibt. Dabei ist aufgefallen, dass manche Mailadressen den Namen erst hinter dem @-Zeichen haben. 

### a.)

Entwickeln Sie bitte einen Python-Ausdruck, der die Namen hinter dem @-Zeichen als Elemente einer Liste ausgibt. 

In [7]:
mail = "service@klara_Klarna.com"

name = mail.split('@')[1].split('.')[0]

namen = name.split('_')

namen


['klara', 'Klarna']

Erwarteter Output:

    ['klara', 'Klarna']

### b.)

Nachdem Sie nun die Namen der AbonnetInnen zu den jeweiligen E-Mail-Adressen zuordnen können, stellt die Chefredakteurin Ihnen eine neue Aufgabe. Überführen Sie bitte die folgende Tabelle in ein Dictionary bestehend aus den Namen als Keys und den E-Mail-Adressen als Values:

| Name      	|   	| E-Mail                          	|
|:-----------	|---	|:---------------------------------	|
| Klara     	|    	| service@klara_Klarna.com        	|
| Sebastian 	|    	| Sebastian_Duesentrieb@gmail.com 	|
| Britta    	|    	| bwiebritta@yahoo.com            	|
| Klaus     	|    	| Klaus@peter.de                  	|

Legen Sie dazu zunächst zwei Listen an, die Sie dann zusammenführen können.

**Tipp:**
<details>

Nutzen Sie die Funktion `enumerate()` um die Indizes und Elemente einer Liste zu verwenden.
</details>

In [4]:
tmpList1 = ['Klara', 'Sebastian', 'Britta', 'Klaus']
tmpList2 =  ["service@klara_Klarna.com", "Sebastian_Duesentrieb@gmail.com", "bwiebritte@yahoo.com", "Klaus@peter.com"]

result = {}
for names, mails in zip(tmpList1, tmpList2):
    result[names] = mails

result



{'Klara': 'service@klara_Klarna.com',
 'Sebastian': 'Sebastian_Duesentrieb@gmail.com',
 'Britta': 'bwiebritte@yahoo.com',
 'Klaus': 'Klaus@peter.com'}

In [5]:
result = {}
for index, value in enumerate(tmpList1):
    print(index)
    print(value)
    result[value] = tmpList2[index]

result

0
Klara
1
Sebastian
2
Britta
3
Klaus


{'Klara': 'service@klara_Klarna.com',
 'Sebastian': 'Sebastian_Duesentrieb@gmail.com',
 'Britta': 'bwiebritte@yahoo.com',
 'Klaus': 'Klaus@peter.com'}

## 3. Anpassung der Werbeeinnahmen

Die Freude im Podcast-Team ist rießig. Zuletzt kamen immer mehr interessierte ZuhörerInnen hinzu und so konnte die Zuhörerzahl in den letzten 3 Monaten um 400% gesteigert werden. Der kaufmännische Leiter des Teams will daher die Einnahmen durch die angebotenen Werbepakete "optimieren". 

Im Moment existieren vier Werbepakete für den Podcast. Bisher wurden immer kundenspezifische Preise bis zu einer Höchstgrenze für jedes Paket verhandelt.

1. Paket "Einleitung", Kosten bis 350 €
2. Paket "Mitte", Kosten bis 300 €
3. Paket "Komplett", Kosten bis 600 €
4. Paket "Zwischendurch", bis 800 €

### a.)

Der kaufmännische Leiter hat beschlossen, dass aufgrund der deutlich vergrößerten Reichweite folgende Preisanpassungen wohl gerechtfertigt sind:
Die Pakete bis 400 € werden um 10% angehoben, die Pakete zwischen 400 und 600 € um 20% und die Pakete über 600 € um 25%.

Bitte setzen Sie die Preisanpassung in einem Pythonskript um, sodass der neue Preis in Abhängigkeit eines variablen Eingabepreises ausgegeben wird.

In [12]:
input_price = 333

if input_price <= 400:
    outputPrice = input_price*1.10
elif input_price <= 600:
    outputPrice =input_price*1.20
else:
    outputPrice = input_price*1.25

print(outputPrice)

366.3


In [17]:
def changePrice(inputPrice):
    if inputPrice <= 400:
        outputPrice = inputPrice * 1.1
    elif inputPrice <= 600:
        outputPrice = inputPrice * 1.2
    else:
        outputPrice = inputPrice * 1.25
    return int(outputPrice)    

# 调用函数
changePrice(333) 

366

Erwarteter Output:

    366.3

### b.)

Durch den Erfolg des Podcasts sind nun auch die Zeiten der kundenspezifischen Preise und Verhandlungen vorbei. Jedes Paket hat ab jetzt den oben genannten Maximalpreis als Festpreis.

Bitte speichern Sie die Paketpreise in eine Liste und generieren Sie mithilfe dieser Liste eine neue Liste, in der die neuen Preis zu finden sind.

In [13]:
oldPrice = [350, 300, 600, 800]
newPrice = []
for price in oldPrice:
    if price <= 400:
        newPrice.append(int(price*1.10))
    elif price <= 600:
        newPrice.append(int(price*1.20))
    else:
        newPrice.append(int(price*1.25))

print(newPrice)

[385, 330, 720, 1000]


In [18]:
oldPrice = [350, 300, 600, 800]
newPrice = [changePrice(x) for x in oldPrice]
newPrice

[385, 330, 720, 1000]

Erwarteter Output:

    [385, 330, 720, 1000]

### c.) 

Nun ist Kassensturz angesagt! Ihr kaufmännischer Leiter will jetzt wissen, was die Preisanpassung in den vergangenen Wochen gebracht hat. 

Dafür soll eine Funktion geschrieben werden, welcher zwei Listen (Preise und Verkaufszahlen) als Parameter übergeben werden und diese Funktion, dann den **Umsatz** zurückgibt.


Hinweis: Sie können annehmen, dass die Indizierung der beiden Listen identisch ist, d.h. der erste Eintrag in der Preisliste gehört zum ersten Eintrag in der Liste der Verkaufszahlen.

In [14]:

sales = [3, 5, 7, 4]
oldPrice = [350, 300, 600, 800]
newPrice = [385, 330, 720, 1000]

def Umsatz(anzahl, preis):
    umsatz = 0
    for index in range(4):
        umsatz = umsatz + anzahl[index] * preis[index]
    return umsatz

print(Umsatz(sales, newPrice) - Umsatz(sales, oldPrice))

def Umsatz(sales, price):

    umsatz = 0

    for index, x in enumerate(sales):
        umsatz = umsatz + x*price[index]
    
    return umsatz

print(Umsatz(sales, newPrice) - Umsatz(sales, oldPrice))

1895


In [22]:
sales = [3, 5, 7, 4]
oldPrice = [350, 300, 600, 800]
newPrice = [385, 330, 720, 1000]

def Turnover(amount, price):
    turnover = 0
    for index in range(len(amount)):
        turnover += price[index] * amount[index]
    return turnover 

print(f'涨价后的广告收入为： {Turnover(sales, newPrice)}')
print(f'涨价前的广告收入为： {Turnover(sales, oldPrice)}')
print(f'涨价前后的差价为： {Turnover(sales, newPrice) - Turnover(sales, oldPrice)}')

涨价后的广告收入为： 11845
涨价前的广告收入为： 9950
涨价前后的差价为： 1895


In [24]:
sales = [3, 5, 7, 4]
oldPrice = [350, 300, 600, 800]
newPrice = [385, 330, 720, 1000]

def Turnover_2(amount, price):
    turnover = 0
    for index, x in enumerate(amount):
        turnover += price[index] * x
    return turnover 

print(f'涨价后的广告收入为： {Turnover_2(sales, newPrice)}')
print(f'涨价前的广告收入为： {Turnover_2(sales, oldPrice)}')
print(f'涨价前后的差价为： {Turnover_2(sales, newPrice) - Turnover(sales, oldPrice)}')

涨价后的广告收入为： 11845
涨价前的广告收入为： 9950
涨价前后的差价为： 1895


Erwarteter Output:

    1895

### d.)

Informieren Sie sich bitte über die Möglichkeiten der lambda-Funktion. Setzen Sie bitte anschließend die Funktion **Umsatz** als lambda-Funktion um.

In [18]:
Umsatz = lambda preis, vkz: sum([preis[i]*vkz[i] for i in range(len(sales))])

print(Umsatz(sales, newPrice) - Umsatz(sales, oldPrice))


1895


In [26]:
Turnover_lambda = lambda price, amount: sum([price[i] * amount[i] for i in range(len(price))])
Turnover_lambda(newPrice, sales) - Turnover_lambda(oldPrice, sales)

1895

Wie können lambda-Funktionen noch eingesetzt werden?

## 4. Expansionswahnsinn

Vom neuerlichen Erfolg des Podcasts getrieben schlägt der kaufmännische Leiter nun vor, mit einem englischsprachigen Ableger des Podcasts in die USA zu expandieren. Die Chefredakteurin ist noch skeptisch und beauftragt Sie mit der Analyse des Marktpotenzials. In der Zielgruppe des Podcasts sind laut Studien die meisten Personen zwischen 25 und 35 Jahren alt.

### a.)

Finden Sie daher heraus, wie viele US-BürgerInnen derzeit in der relevanten Altergruppe sind (einschließlich der genannten Grenzen). Leider hat die Chefredakteurin nur eine Tabelle auftreiben können, die alle Geburten nach Bundesstaaten und Vornamen gliedert. Versuchen Sie die benötigten Informationen aus diesem Datensatz zu generieren(births.csv)!

In [29]:
import pandas as pd
data = pd.read_csv('births.csv')

Erwarteter Output:

    36976366

In [30]:
import pandas as pd

df_births = pd.read_csv('births.csv')

df_births.Count[(2021 - df_births.Year <= 35) & (2021 - df_births.Year >= 25)].sum()

36976366

In [50]:
df_births.head()
df_births['Age'] = 2021 - df_births['Year'] 
df_births.head()

state Gender  Year   Name  Count  Age
0    IN      F  1961   Lisa   1469   60
1    IN      F  1961   Mary   1151   60
2    IN      F  1961  Karen    941   60
3    IN      F  1961  Susan    941   60
4    IN      F  1961  Linda    919   60

In [53]:
# 增加年龄列
df_births['Age'] = 2021 - df_births['Year'] 
df_births = df_births[(df_births['Age'] >= 25) & (df_births['Age'] <= 35)]

# 按照年龄分组，并计算每个年龄段人数的总和
age_count = df_births.groupby('Age')['Count'].sum()


# 创建字典
age_count_dict = dict(age_count)

# 获取字典中对应最大的值的键
max_key = max(age_count_dict, key = age_count_dict.get)
print(max_key)
print(age_count_dict[max_key])


31
3567577


C:\Users\Heinrich\AppData\Local\Temp\ipykernel_17288\1110181520.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_births['Age'] = 2021 - df_births['Year']


In [37]:
age_count

Age
2     2891772
3     2949060
4     3008598
5     3098778
6     3134011
7     3141425
8     3079348
9     3083008
10    3086059
11    3122128
12    3236955
13    3344482
14    3416158
15    3390724
16    3307852
17    3296174
18    3290969
19    3239855
20    3252723
21    3298520
22    3234596
23    3230662
24    3194233
25    3225793
26    3249949
27    3307859
28    3364630
29    3442904
30    3503407
31    3567577
32    3475994
33    3342082
34    3267158
35    3229013
36    3246990
37    3177582
38    3154303
39    3193594
40    3146595
41    3130961
42    3019120
43    2877643
44    2882873
45    2753960
46    2744584
47    2773958
48    2759369
49    2890095
50    3177630
51    3359449
52    3244912
53    3159689
54    3184310
55    3267847
56    3419530
57    3674740
58    3747748
59    3825785
60    3930597
Name: Count, dtype: int64

#### b.)

Die Chef-Redakteurin ist von Ihrem Ergebnis angetan. Um das Programm passgenau auf die meisten potenziellen ZuhörerInnen abzustimmen, bittet sie Sie um eine weitere Analyse. 

Bitte finden Sie heraus, welches Alter die meisten Personen innerhalb der relevanten Altersgruppe haben. Nutzen Sie dabei ein Dictionary, welches als Schlüssel das Alter und als Wert die jeweilige Personenanzahl beinhaltet.

In [ ]:
with open('births.csv', "r") as csvFile:
    myDict = {}

    for line in csvFile:
        split_line = line.strip().split(',') # strip用于移除每行首位的空格

        if split_line[2] =='Year': # 这个if语句只有第一行符合，然后执行continue，跳过第一轮循环进入第二轮
            continue

        year_lb = 2021 - 35     # 规定年龄组的上下限
        year_ub = 2021 - 25

        if int(split_line[2]) >= year_lb and int(split_line[2]) <= year_ub:

            age = 2021 - int(split_line[2])

            counter = int(split_line[4])

            if str(age) in myDict.keys():  # 这个结构很普遍，记住！适用于在字典里设置计数器
                myDict[str(age)] += counter
            
            else:
                myDict[str(age)] = counter

max_ag = 0
age = ""
for key, value in myDict.items():
    if max_ag < value:
        max_ag = value
        age = key

print(age)
print(max_ag)

Erwarteter Output:

    31
    3567577